File: benchmarks/swift/string_processing.swift
// String processing benchmark for Swift

import Foundation

// Create a random string of specified length
func createRandomString(length: Int) -> String {
    let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    return String((0..<length).map { _ in chars.randomElement()! })
}

// Process a string with various operations
func processString(_ input: String) -> String {
    // Replace all vowels with uppercase version
    var result = input
        .replacingOccurrences(of: "a", with: "A")
        .replacingOccurrences(of: "e", with: "E")
        .replacingOccurrences(of: "i", with: "I")
        .replacingOccurrences(of: "o", with: "O")
        .replacingOccurrences(of: "u", with: "U")
    
    // Replace digits with doubled value
    for i in 0...9 {
        result = result.replacingOccurrences(of: String(i), with: String(i * 2))
    }
    
    // Capitalize first letter if string is not empty
    if let firstChar = result.first {
        result = String(firstChar).uppercased() + result.dropFirst()
    }
    
    // Reverse the string
    let reversed = String(result.reversed())
    
    // Take first half
    let halfLength = reversed.count / 2
    return String(reversed.prefix(halfLength))
}

// Process multiple strings of different sizes
func processStrings(count: Int, size: Int) -> String {
    var result = ""
    
    for _ in 0..<count {
        let str = createRandomString(length: size)
        let processed = processString(str)
        result += processed
    }
    
    return result
}

func main() {
    let startTime = Date()
    
    // Process strings of different sizes
    let small = processStrings(count: 10000, size: 10)     // 10,000 strings of length 10
    let medium = processStrings(count: 1000, size: 100)    // 1,000 strings of length 100
    let large = processStrings(count: 100, size: 1000)     // 100 strings of length 1,000
    
    let resultLength = small.count + medium.count + large.count
    print("Processed string length: \(resultLength)")
    
    let elapsedTime = -startTime.timeIntervalSinceNow * 1000
    print("Time taken: \(elapsedTime) ms")
}

main()

File: benchmarks/haskell/string_processing.hs
-- String processing benchmark for Haskell

import System.Environment
import System.IO
import Control.Monad
import Data.Time.Clock
import Text.Printf
import Data.Char
import System.Random
import qualified Data.Text as T

-- Create a random string of specified length
createRandomString :: Int -> IO String
createRandomString length = do
    let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    indices <- replicateM length $ randomRIO (0, length chars - 1)
    return $ map (chars !!) indices

-- Process a string with various operations
processString :: String -> String
processString input = do
    -- Replace all vowels with uppercase version
    let replaceVowels c = case c of
            'a' -> 'A'
            'e' -> 'E'
            'i' -> 'I'
            'o' -> 'O'
            'u' -> 'U'
            _ -> c
    
    let replaced = map replaceVowels input
    
    -- Replace digits with doubled value
    let replaceDigits c = if isDigit c
                          then let digit = digitToInt c
                               in head (show (digit * 2))
                          else c
    
    let replacedDigits = map replaceDigits replaced
    
    -- Capitalize first letter if string is not empty
    let capitalized = case replacedDigits of
            [] -> []
            (c:cs) -> toUpper c : cs
    
    -- Reverse the string
    let reversed = reverse capitalized
    
    -- Take first half
    take (length reversed `div` 2) reversed

-- Process multiple strings of different sizes
processStrings :: Int -> Int -> IO String
processStrings count size = do
    strings <- replicateM count $ createRandomString size
    return $ concatMap processString strings

-- Main function
main :: IO ()
main = do
    startTime <- getCurrentTime
    
    -- Process strings of different sizes
    small <- processStrings 10000 10    -- 10,000 strings of length 10
    medium <- processStrings 1000 100   -- 1,000 strings of length 100
    large <- processStrings 100 1000    -- 100 strings of length 1,000
    
    let resultLength = length small + length medium + length large
    printf "Processed string length: %d\n" resultLength
    
    endTime <- getCurrentTime
    let diff = diffUTCTime endTime startTime
    printf "Time taken: %s ms\n" (show (diff * 1000))